File: snuba/utils/metrics/clock.py
import time
from abc import ABC, abstractmethod


class Clock(ABC):
    @abstractmethod
    def time(self) -> float:
        raise NotImplementedError

    @abstractmethod
    def sleep(self, duration: float) -> None:
        raise NotImplementedError


class SystemClock(Clock):
    def time(self) -> float:
        return time.time()

    def sleep(self, duration: float) -> None:
        time.sleep(duration)


class TestingClock(Clock):
    def __init__(self) -> None:
        self.__time = 0.0

    def time(self) -> float:
        return self.__time

    def sleep(self, duration: float) -> None:
        self.__time = self.__time + duration

File: rust_snuba/rust_arroyo/src/utils/clock.rs
use std::thread::sleep;
use std::time::{Duration, SystemTime};

pub trait Clock: Send {
    fn time(&self) -> SystemTime;

    fn sleep(&mut self, duration: Duration);
}

pub struct SystemClock {}
impl Clock for SystemClock {
    fn time(&self) -> SystemTime {
        SystemTime::now()
    }

    fn sleep(&mut self, duration: Duration) {
        sleep(duration)
    }
}

pub struct TestingClock {
    time: SystemTime,
}

impl TestingClock {
    pub fn new(now: SystemTime) -> Self {
        Self { time: now }
    }
}

impl Clock for TestingClock {
    fn time(&self) -> SystemTime {
        self.time
    }

    fn sleep(&mut self, duration: Duration) {
        self.time += duration;
    }
}
